Hummingbird.docc/Articles/Recommendations.md

```diff
@@ -0,0 +1,342 @@
+
+```swift
+let router = Router(context: MyContext.self)
+router.add("users", routes: UserController().routes)
+```
+
+## Request Context
+
+It's also recommended to users that they refine the AppContext requirements for only as much as strictly required by the routes. For example:
+
+```swift
+// ✅ This is recommended, because it allows a different RequestContext to be used for login routes
+struct UserController<Context: RequestContext> { ... }
+```
+
+If a route has requirements for the request, like being authenticated, it's recommended to leverage _protocol composition_ to refine the RequestContext.
+
+```swift
+// ✅ This is recommended, because it leverages protocol composition to refine the RequestContext
+// while still being flexible enough to allow different concrete types to fulfil the requirements
+struct UserProfileController<
+    Context: RequestContext & AuthenticatedRequestContext
+> { ... }
+```
+
+Users should define various traits related to a request's state (context) in separate protocols.
+
+```swift
+protocol RateLimitedRequestContext {
+    var rateLimitStatus: RateLimitStatus { get }
+    var ipAddress: String { get }
+}
+
+protocol AuthenticatedRequestContext {
+    var user: User { get }
+}
+```
+
+## Middleware
+
+Middleware follow the same RequestContext conventions and recommendations as Routes. There's one distinction between Routes in that middleware can modify the request context.
+
+```swift
+public struct AuthenticationGuardMiddleware<Context: AuthRequestContext>: RouterMiddleware {
+    public init() {}
+
+    public func handle(_ request: Request, context: Context, next: (Request, Context) async throws -> Response) async throws -> Response {
+        guard context.identity != nil else {
+            throw HTTPError(.unauthorized)
+        }
+        return try await next(request, context)
+    }
+}
+```
+
+### Child Request Contexts
+
+```swift
+struct MyAppRequestContext: RequestContext {
+    public var coreContext: CoreRequestContextStorage
+    // User is optional, because it's not required for all routes
+    public var user: User?
+
+    public init(source: Source) {
+        self.coreContext = .init(source: source)
+        self.routerContext = .init()
+        self.user = nil
+    }
+}
+```
+
+```swift
+struct AuthenticatedRequestContext: ChildRequestContext {
+    public var coreContext: CoreRequestContextStorage
+    // User is required in all routes, and this is statically known
+    public var user: User
+
+    init(context: MyAppRequestContext) throws {
+        guard let user = context.user else {
+            throw HTTPError(.unauthorized)
+        }
+
+        self.coreContext = context.coreContext
+        self.user = user
+    }
+}
+```
+
+## Services
+
+Users should define (external) resources in `Services`. For example:
+
+- PostgreSQL
+- S3 Bucket Management
+- Job Queues
+- Caching
+- OpenTelemetry
+- File System Access
```

Comment:
Some of these don't actually require any additional code as they come from other libraries. Also I'm wondering if we should add a section describing each folder in the project structure section of the doc. The services section should be reserved for talking about conforming to ServiceLifecycle